File: src/sota_ops/ball_query.cpp
#include <torch/library.h>
#include "sota_ops/ball_query.h"

namespace sota_ops::ball_query {

std::tuple<at::Tensor, at::Tensor> ball_query(
    at::Tensor points,
    at::Tensor query,
    at::Tensor batch_indices,
    at::Tensor batch_offsets,
    double radius,
    int64_t num_samples) {
  static auto op = c10::Dispatcher::singleton()
                       .findSchemaOrThrow("sota_ops::ball_query", "")
                       .typed<decltype(ball_query)>();
  return op.call(points, query, batch_indices, batch_offsets, radius, num_samples);
}

TORCH_LIBRARY_FRAGMENT(sota_ops, m) {
  m.def(TORCH_SELECTIVE_SCHEMA(
      "sota_ops::ball_query(Tensor points, Tensor query, Tensor batch_indices, "
      "Tensor batch_offsets, float radius, int num_samples) -> (Tensor, Tensor)"));
}

} // namespace sota_ops::ball_query

File: sota_ops/ball_query.py
from typing import Tuple, Optional

import torch


@torch.no_grad()
def ball_query(
    points: torch.Tensor,
    query: torch.Tensor,
    batch_indices: torch.Tensor,
    batch_offsets: torch.Tensor,
    radius: float,
    num_samples: int,
    point_labels: Optional[torch.Tensor] = None,
    query_labels: Optional[torch.Tensor] = None,
) -> Tuple[torch.Tensor, torch.Tensor]:
    points = points.contiguous()
    query = query.contiguous()
    batch_indices = batch_indices.contiguous()
    batch_offsets = batch_offsets.contiguous()

    if point_labels is not None:
        point_labels = point_labels.contiguous()

    if query_labels is not None:
        query_labels = query_labels.contiguous()

    return torch.ops.sota_ops.ball_query(
        points, query, batch_indices, batch_offsets, radius, num_samples,
        point_labels, query_labels,
    )
